### PR TITLE
feat: migrate packagist to v2

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes:

1. Updated code to v2 for packagist.org.
2. Updated test cases and snapshots to use v2 URL. 

## Context:

- Ref : #13359 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/packagist/index.ts
Comment (viceice) [inline lib/datasource/packagist/index.ts:205]: Please use `joinUrlParts` from `utils/url`.

This is all we need to change? No result object changes?

Then we don't need to change the cache key, as they seem to be compatible.

Comment (RahulGautamSingh) [reply lib/datasource/packagist/index.ts:205]: I could only see 3 changes in the new API.
1. In the URL `p2` instead of `p`. In the docs, the URL used is
` https://repo.packagist.org/p2/monolog/monolog.json`, I didn't change the URL since both formats work fine.

2. The versions list is returned in an array instead of an object.
-> Due to the new return object containing an array, it seemed like we might have to change the logic for `extractDepReleases` but the way the fn is written handles the array as well.

3. It is in a compressed format for efficiency which requires you to use [link](https://github.com/composer/metadata-minifier/blob/main/src/MetadataMinifier.php#L17) from the package to restore it to the full data.
-> After comparing the returned object from old and new API, info like name, source, description..were being repeated in each array element in the older version, in the new API and this info is only found in the first element of the `packages` array.

4. The p2/$vendor/$package.json file contains only tagged releases. If you want to fetch information about branches (i.e. dev versions) you need to download p2/$vendor/$package~dev.json.
-> Want to discuss this one. In the old code, we are storing the dev versions as well. To Implement we fetch data and push it to the result objects `packages[nameOfDep]` field.

File: lib/datasource/packagist/index.ts
Comment (viceice) [inline lib/datasource/packagist/index.ts:205]:
```suggestion
  const pkgUrl = joinUrlParts(regUrl, `/p2/${name}.json`);
  // TODO: fix types (#9610)
```
Don't remove the comment, as you don't fixed it😉

Comment (RahulGautamSingh) [reply lib/datasource/packagist/index.ts:205]: Removed by mistake 😅

**Comment by RahulGautamSingh** [comment]
> > I have listed the 4 changes in the new API and how they have been handled below.
> > 
> > 1. **In the URL `p2` instead of `p`.**
> > 
> > > In the docs, the URL used is ` https://repo.packagist.org/p2/monolog/monolog.json`, I didn't change the URL since both formats work fine as this option was available for old API as well and we were not using it.
> > 
> > 
> > 
> > 2. **The versions list is returned in an array instead of an object.**
> > 
> > > Due to the new return object containing an array, it seemed like we might have to change the logic for `extractDepReleases` but the way the fn is written handles the array as well.
> > 
> > 
> > 
> > 3. _It is in a compressed format for efficiency which requires you to use [link](https://github.com/composer/metadata-minifier/blob/main/src/MetadataMinifier.php#L17) from the package to restore it to the full data._
> > 
> > > After comparing the returned object from old and new API, info like name, source, description..were being repeated in each array element in the older version, in the new API and this info is only found in the first element of the `packages` array.
> > 
> > 
> > 
> > 4. **The p2/$vendor/$package.json file contains only tagged releases. If you want to fetch information about branches (i.e. dev versions) you need to download p2/$vendor/$package~dev.json.**
> > 
> > > Want to discuss this one. In the old code, we are storing the dev versions as well. To Implement we fetch data and push it to the result objects `packages[nameOfDep]` field.
> 
> We need to make sure existing code doesn't break. So when changing the api url is enough to get it done, well i'm fine. But i think we need to update the test, so the request returns an object in API v2 format.

Should we store the dev-releases too? {Point 4}

**Comment by RahulGautamSingh** [comment]
Do we need to store the dev-releases too?

**Comment by JamieMagee** [comment]
@RahulGautamSingh do you mind holding off merging this until I get #12227 merged? I have a lot more changes in that PR and fixing merge conflicts will be more difficult.

**Comment by viceice** [comment]
moving to draft to not accidently merge 😉

**Comment by viceice** [comment]
> Do we need to store the dev-releases too?

Not sure, but i would say yes, so users are able to get unstable updates with [`ignoreUnstable=true`](https://docs.renovatebot.com/configuration-options/#ignoreunstable)

**Comment by rarkins** [comment]
Datasources should usually return all possible releases and let the versioning scheme do the filtering afterwards

**Review by viceice — COMMENTED** [review]
Needs some cleanup in a later PR. I like to see some real repo tests for packagist.org and a v1 custom registry.

**Comment by viceice** [comment]
Can you pin the drupal does and reduce to 2 deps only. So we get h one dep from packagist and one from drupal registry?

**Comment by viceice** [comment]
Looks like drupal dep is using a range, it should be pinned too, then you should see an update when you lower the patch version.

**Review by viceice — APPROVED** [review]

**Pull request merged**
